### PR TITLE
BSE-5119: torch trainer

### PR DESCRIPTION
**Pull request by IsaacWarren**
## Changes included in this PR
As title, includes the new torch trainer, model preparation and docs.

## Testing strategy
Unit test

## User facing changes
New functionality

## Checklist
- [x] Pipelines passed before requesting review. To run CI you must include `[run CI]` in your commit message.
- [x] I am familiar with the [Contributing Guide](https://github.com/bodo-ai/Bodo/blob/main/CONTRIBUTING.md) 
- [x] I have installed + ran pre-commit hooks.

### CODE REVIEW

File: bodo/libs/distributed_api.py [hunk 1]
```diff
@@ -2792,7 +2792,7 @@ def get_num_gpus(framework="torch"):  # pragma: no cover
         try:
             import torch
 
-            return torch.cuda.device_count()
+            return torch.accelerator.device_count()
```
Comment (scott-routledge2) [inline bodo/libs/distributed_api.py]: accelerator requires torch>2.5

File: bodo/tests/test_df_lib/test_ai_funcs.py [hunk 2]
```diff
@@ -390,3 +391,78 @@ def test_embed_bedrock_default_formatter():
...
+                self.linear1 = nn.Linear(2, 32)
+                self.relu = nn.ReLU()
+                self.linear2 = nn.Linear(32, 1)
+
+            def forward(self, x):
+                return self.linear2(self.relu(self.linear1(x)))
+
+        model = SimpleModel()
+        model = bodo.ai.train.prepare_model(model, parallel_strategy="ddp")
+        gpu_ranks = bodo.get_gpu_ranks()
```
Comment (scott-routledge2) [inline bodo/tests/test_df_lib/test_ai_funcs.py:426]: Would the user need to add this rebalancing code themselves? Could we potentially do this for them or provide a function similar to `ray.train.torch.prepare_data_loader()`

File: docs/docs/api_docs/dataframe_lib/ai.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        # If we're using an accelerator, rebalance data to match GPU ranks
+        data = bodo.rebalance(data, dests=gpu_ranks)
+
+    # Train on data
+    criterion = nn.MSELoss()
+    optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
+    for epoch in range(config.get("epochs", 5)):
+        batch_size = config.get("batch_size", 2)
+        for i in range(0, len(data), batch_size):
+            batch = data[i : i + batch_size]
```
Comment (scott-routledge2) [inline docs/docs/api_docs/dataframe_lib/ai.md:105]: I was wondering if we should create some Dataset sub-classes to simplify the training code that needs to be written a bit. If the subclass doesn't fit the users needs they can always create their own subclass or write the training loop using raw dataframes

For example for a supervised learning task:
``` py
class DataFrameDataset(Dataset):
    def __init__(self, dataframe, label_col, features_col):
        self.dataframe = dataframe
        self.label_col = label_col
        self.feature_col = features_col

    def __len__(self):
        return len(self.dataframe)

    def __getitem__(self, idx):
        label = self.dataframe.iloc[idx][self.label_col]
        features = self.dataframe.iloc[idx][self.feature_col]
        return torch.tensor(features), torch.tensor(label)

dataset = DataFrameDataset(df, label_col="label", features_col="features")
dataloader = DataLoader(dataset, batch_size=2, shuffle=True)

for X,y in dataloader:
    # training code here
```

File: bodo/__init__.py [hunk 4]
```diff
@@ -380,6 +380,12 @@ def random_shuffle(*args, **kwargs):
     from bodo.libs.distributed_api import random_shuffle
     return random_shuffle(*args, **kwargs)
 
+def get_gpu_ranks(*args, **kwargs):
```
Comment (scott-routledge2) [inline bodo/__init__.py:389]: If this is only called on workers then it might cause hangs if the compiler wasn't already imported on the spawner

File: bodo/tests/test_df_lib/test_ai_funcs.py [hunk 5]
```diff
@@ -390,3 +391,78 @@ def test_embed_bedrock_default_formatter():
...
+        model_device = next(model.parameters()).device
+        if model_device.type != "cpu":
+            # If we're using an accelerator, rebalance data to match GPU ranks
+            data = bodo.rebalance(data, dests=gpu_ranks)
+
+        # train on data
+        criterion = nn.MSELoss()
+        optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
+        for epoch in range(config.get("epochs", 5)):
+            batch_size = config.get("batch_size", 2)
```
Comment (scott-routledge2) [inline bodo/tests/test_df_lib/test_ai_funcs.py:440]: Wouldn't this cause hangs if the batch size is not evenly divisible by the number of ranks (so some ranks execute more batches than others)?

File: docs/docs/api_docs/dataframe_lib/ai.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+    model_device = next(model.parameters()).device
+    if model_device.type != "cpu":
+        # If we're using an accelerator, rebalance data to match GPU ranks
+        data = bodo.rebalance(data, dests=gpu_ranks)
+
+    # Train on data
+    criterion = nn.MSELoss()
+    optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
+    for epoch in range(config.get("epochs", 5)):
```
Comment (scott-routledge2) [inline docs/docs/api_docs/dataframe_lib/ai.md:102]: I think the semantics of this training loop are different than vanilla FSDP since we are not doing any shuffling, and data samples always map to the same rank. Maybe we need our own version of DistributedSampler? https://github.com/pytorch/pytorch/blob/ba56102387ef21a3b04b357e5b183d48f0afefc7/torch/utils/data/distributed.py#L107

**Review by scott-routledge2 — APPROVED** [review]
Thanks Isaac! Overall looks good to me.

File: bodo/ai/train.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+def _get_open_port():
+    """
+    Finds and returns an available open TCP port.
+    """
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    try:
+        s.bind(("", 0))  # Bind to an ephemeral port (port 0)
+        port = s.getsockname()[1]  # Get the assigned port number
+        return port
```
Comment (DrTodd13) [inline bodo/ai/train.py:33]: Is this safe?  There's a time window after closing the socket that some other process could open that port number.

File: bodo/tests/test_df_lib/test_ai_funcs.py [hunk 8]
```diff
@@ -390,3 +391,78 @@ def test_embed_bedrock_default_formatter():
...
+                model.module
+                if isinstance(model, torch.nn.parallel.DistributedDataParallel)
+                else model
+            )
+            torch.distributed.checkpoint.state_dict_saver.save(
+                {"model_state_dict": base_model.state_dict()},
+                checkpoint_id=config["checkpoint_dir"],
+            )
+
+            print(f"Epoch {epoch}, Loss: {loss.item()}")
```
Comment (DrTodd13) [inline bodo/tests/test_df_lib/test_ai_funcs.py]: Is this needed?

**Review by DrTodd13 — APPROVED** [review]
Thanks.

File: bodo/ai/train.py (hunk 7 above)
Comment (IsaacWarren) [reply bodo/ai/train.py:33]: That's true but I think it's better than what pytorch recommends which is just a hardcoded port assignment. I'll wrap in a retry

File: bodo/__init__.py (hunk 4 above)
Comment (IsaacWarren) [reply bodo/__init__.py:389]: Why? Do we have a solution?

Comment (scott-routledge2) [reply bodo/__init__.py:389]: When the compiler is imported on the workers, `scatterv` gets replaced with `bodo.libs.distributed_api.scatterv`. The next time someone goes to scatter a dataframe, the spawner will call `scatterv_nojit` and the workers call `bodo.libs.distributed_api.scatterv`, which hangs. One solution could be to import the compiler before `submit_func_to_workers` since training is not going to be very responsive anyways.
